R/derive_vars_trt_start.R

```diff
@@ -0,0 +1,158 @@
+#'
+#' A list of dates is expected. It is ensured that the imputed date is not after
+#' any of the specified dates, e.g., that the imputed date is not after the data
+#' cut off date. Only dates which are in the range of possible dates are
+#' considered.
+#'
+#' @param ord_vars
+#'  Order of variables for which either the last or first observation
+#'  is picked as treatment start date
+#'  e.g. vars(EXSTDTC,EXSEQ)
+#'
+#' @param ord_filter
+#'  The mode by which the filter_extreme is chosen.
+#'  If "first", the first observation of each by group is included in the output dataset.
+#'  If "last", the last observation of each by group is included in the output dataset.
+#'
+#'  Permitted Values: "first", "last"
+#'
+#' @details For each group (with respect to the variables specified for the
+#'   `by_vars` parameter) the first observation (with respect to the order
+#'   specified for the `order` parameter) is included in the output dataset.
+#'
+#' @author Stefan Bundfuss, Teckla Akinyi
+#'
+#' @return The input dataset with start date variable added e.g. `TRTSDTM`
+#'
+#' @export
+#'
+#' @keywords adsl timing derivation
+#'
+#' @examples
+#' library(dplyr, warn.conflicts = FALSE)
+#' data("ex")
+#' data("dm")
+#'
+#' dm %>%
+#'   derive_vars_trt_start(dataset_ex = ex
+#'                         filter_ex =  (EXDOSE > 0 | (EXDOSE == 0 & str_detect(EXTRT, "PLACEBO"))) & nchar(EXSTDTC) >= 10, # nolint
+#'                         subject_keys = vars(STUDYID, USUBJID),
+#'                         new_var=TRTSDTM,
+#'                         dtc=EXSTDTC,
+#'                         date_imputation = "first",
+#'                         time_imputation="first",
+#'                         min_dates = NULL,
+#'                         max_dates = NULL,
+#'                         ord_vars = vars(EXSTDTC,EXSEQ),
+#'                         ord_filter="first") %>%
+#'   select(USUBJID, TRTSDTM)
+derive_vars_trt_start <- function(dataset,
+                                 dataset_ex,
+                                 filter_ex = NULL,
+                                 subject_keys,
+                                 new_var,
+                                 dtc,
+                                 date_imputation,
+                                 time_imputation,
+                                 min_dates = NULL,
+                                 max_dates = NULL,
+                                 ord_vars,
+                                 ord_filter) {
+    new_var <- assert_symbol(enquo(new_var))
+    dtc <- assert_symbol(enquo(dtc))
+    ord <- assert_symbol(enquo(ord))
+    assert_data_frame(dataset,subject_keys)
+    assert_data_frame(dataset_ex, required_vars = quo_c(subject_keys, ord_vars))
+    assert_character_scalar(ord_filter, values = c("first","last"))
+
+    filter_ex <- assert_filter_cond(enquo(filter_ex), optional = TRUE)
+
+    add <- dataset_ex %>%
+      mutate(imputed_dtc = convert_dtc_to_dtm(!!dtc,
```

Comment:
Do we also want the option to derive a `--SDT`? If not, should we rename the function so that it is clear that it returns the `--SDTM`? Or make it very clear in the documentation?

Reply:
I believe the original was --DTM?

Reply:
Yes, I agree. So we should make it clear in the documentation because it is not obvious from the function name (i think ;0). 
Also, we should probably add a warning if the user enter a `new_var` like `TRTSDT` rather than `TRTSDTM`?